plutus-chain-index/app/Marconi/Logging.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE NamedFieldPuns    #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Marconi.Logging (logging) where
+
+import Cardano.Api (Block (Block), BlockHeader (BlockHeader), BlockInMode (BlockInMode), CardanoMode,
+                    ChainPoint (ChainPoint), ChainTip (ChainTip), SlotNo (SlotNo, unSlotNo))
+import Cardano.BM.Trace (Trace, logInfo)
+import Control.Monad (when)
+import Data.IORef (IORef, modifyIORef', newIORef, readIORef)
+import Data.Text (Text)
+import Data.Time (NominalDiffTime, UTCTime, defaultTimeLocale, diffUTCTime, formatTime, getCurrentTime,
+                  getCurrentTimeZone, utcToZonedTime)
+import Marconi.Orphans ()
+import Plutus.Streaming (ChainSyncEvent (RollBackward, RollForward))
+import Prettyprinter (Pretty (pretty), defaultLayoutOptions, layoutPretty, (<+>))
+import Prettyprinter.Render.Text (renderStrict)
+import Streaming (Of, Stream, effect)
+import Streaming.Prelude qualified as S
+import Text.Printf (printf)
+
+data SyncStats = SyncStats
+  { -- | Number of applied blocks since last message
+    syncStatsAppliedBlocks :: !Int,
+    -- | Number of rollbacks
+    syncStatsLastRollback  :: !(Maybe UTCTime),
+    -- | Timestamp of last printed message
+    syncStatsLastMessage   :: !(Maybe UTCTime)
+  }
+
+data SyncState = NotSynchronising | Synchronising SlotNo SlotNo Double | Synchronised
```

Comment:
Can you add haddock comments on what each means. Especially, what's the 3 params of `Synchronising`?

Reply:
Will fix

Reply:
Better still, that datatype was used only once so I inlined the case statement and got rid of it.